### PR TITLE
[SQL-DS-CACHE-141][POAE7-1147] add compression/decompression in disaggrated APE

### PR DESCRIPTION
**Pull request by yma11**
## What changes were proposed in this pull request?

add zstd compression/decompression in disaggrated APE

## How was this patch tested?
q41

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by yma11** [comment]
@jikunshang @iyupeng Please take a review. Thanks.

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java [hunk 1]
```diff
@@ -610,42 +615,79 @@ void write(ChannelOutboundInvoker out, ChannelPromise promise, ByteBufAllocator
...
-            // write to channel
-            out.write(headerBuf);
-            out.write(compositedElementLengths);
-            for (int i = 0; i < columnCount; i++) {
-                out.write(dataBuffers[i]);
-            }
-            if (columnCount > 0) {
+            originLength += (columnCount * rowCount); // for null buffers
+            if (!compressEnabled) {
+                // set header
```
Comment (jikunshang) [inline oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: possible to refactor this part as a method?

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/util/ZStdUtils.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package com.intel.ape.util;
```
Comment (jikunshang) [inline oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/util/ZStdUtils.java:18]: add license please. And I noticed that zstd-jni is using BSD license, any risk? @winningsix  
https://github.com/luben/zstd-jni#license

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java [hunk 3]
```diff
@@ -610,42 +615,79 @@ void write(ChannelOutboundInvoker out, ChannelPromise promise, ByteBufAllocator
...
+                CompositeByteBuf compositeByteBuf = Unpooled.compositeBuffer(2 * columnCount + 1);
+                compositeByteBuf.addComponent(compositedElementLengths);
+                for (int i = 0; i < columnCount; i++) {
+                    compositeByteBuf.addComponent(dataBuffers[i]);
+                }
+                for (int i = 0; i < columnCount; i++) {
+                    compositeByteBuf.addComponent(nullBuffers[i]);
+                }
+                try {
+                    ByteBuf compressedData = ZStdUtils.compress(compositeByteBuf);
```
Comment (jikunshang) [inline oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: It's ok to directly use `ZStdUtils` here. But in the future we may support more kind of codec, user could set codec type in profile. Let's do a refactor in later PRs.

**Review by jikunshang — COMMENTED** [review]
Overall LGTM.

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java [hunk 4]
```diff
@@ -610,44 +615,85 @@ void write(ChannelOutboundInvoker out, ChannelPromise promise, ByteBufAllocator
...
```
Comment (iyupeng) [inline oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: `compositeByteBuf` may need a release.

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java (hunk 1 above)
Comment (jikunshang) [reply oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: Maybe I didn't make it clear. What I expected is something like
`if(compressEnabled) {
    doCompressThings();
} else {
   doNonCompressThings(); 
}`

File: oap-ape/ape-java/ape-common/src/test/java/com/intel/ape/util/ZStdUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package com.intel.ape.util;
```
Comment (jikunshang) [inline oap-ape/ape-java/ape-common/src/test/java/com/intel/ape/util/ZStdUtilsTest.java:18]: license.

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java (hunk 3 above)
Comment (yma11) [reply oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: Yes. sure.

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java [hunk 6]
```diff
@@ -666,19 +712,26 @@ static ReadBatchResponse readFrom(ByteBuf buffer) {
...
-            for (int i = 0; i < columnCount; i++) {
-                dataBuffers[i] = buffer.readRetainedSlice(dataBufferLengths[i]);
-            }
-
             ByteBuf[] nullBuffers = new ByteBuf[columnCount];
-            for (int i = 0; i < columnCount; i++) {
-                nullBuffers[i] = buffer.readRetainedSlice(rowCount);
+            try {
+                ByteBuf content = compressEnabled ?
+                        ZStdUtils.decompress(buffer.retainedSlice(), contentLength)
```
Comment (iyupeng) [inline oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: The first `buffer.retainedSlice()`  and `content` may need `release`.

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java (hunk 1 above)
Comment (yma11) [reply oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: updated.

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java [hunk 7]
```diff
@@ -633,20 +655,62 @@ void write(ChannelOutboundInvoker out, ChannelPromise promise, ByteBufAllocator
...
+            // track content length which will be used for decompress
+            headerBuf.writeBoolean(false);
+            headerBuf.writeInt(originLength);
             out.write(headerBuf);
+            // write content to channel
             out.write(compositedElementLengths);
             for (int i = 0; i < columnCount; i++) {
                 out.write(dataBuffers[i]);
             }
-            if (columnCount > 0) {
```
Comment (iyupeng) [inline oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java:643]: should keep this. `columnCount` could be 0

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java [hunk 8]
```diff
@@ -666,19 +730,28 @@ static ReadBatchResponse readFrom(ByteBuf buffer) {
...
-            for (int i = 0; i < columnCount; i++) {
-                dataBuffers[i] = buffer.readRetainedSlice(dataBufferLengths[i]);
-            }
-
             ByteBuf[] nullBuffers = new ByteBuf[columnCount];
-            for (int i = 0; i < columnCount; i++) {
-                nullBuffers[i] = buffer.readRetainedSlice(rowCount);
+            try {
+                ByteBuf content = compressEnabled ?
+                        ZStdUtils.decompress(buffer.retainedSlice(), contentLength)
```
Comment (iyupeng) [inline oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: buffer created by `buffer.retainedSlice()` does not get released in after decompression.

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java [hunk 9]
```diff
@@ -610,18 +615,35 @@ void write(ChannelOutboundInvoker out, ChannelPromise promise, ByteBufAllocator
...
             // columnCount(4), rowCount(4), dataBufferLengths(columnCount * 4),
-            // compositeFlags(columnCount), compositedElementCount(4)
-            int headerLength = 4 + 1 + 4 + 4 + columnCount * 4 + columnCount + 4;
-            int contentLength = 0;
-            contentLength += compositedElementLengths.readableBytes();
+            // compositeFlags(columnCount), compositedElementCount(4), compressEnabled(1),
+            // contentLength(4)
+            int headerLength = 4 + 1 + 4 + 4 + columnCount * 4 + columnCount + 4 + 1 + 4;
+            int originLength = getContentLength();
+            if (!compressEnabled) {
```
Comment (iyupeng) [inline oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java:622]: if column count == 0, we don't need compression too.

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java (hunk 8 above)
Comment (iyupeng) [inline oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: avoid decompression when content length == 0

Comment (yma11) [reply oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: I did that in decompress method.

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java (hunk 7 above)
Comment (yma11) [reply oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java:643]: So if columnCount =0, we just need to write the headerBuf? no need to write compositedElementLengths as well as dataBuffers?

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java (hunk 8 above)
Comment (iyupeng) [reply oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java]: ok. I didn't notice that

File: oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java (hunk 7 above)
Comment (iyupeng) [reply oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java:643]: yes.

Comment (yma11) [reply oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java:643]: Refact the code. Please take review again.

**Review by jikunshang — APPROVED** [review]

**Review by iyupeng — APPROVED** [review]

**Pull request merged**